File: core-dotnet/about/AboutNode.cs
namespace Core.about;

/// <summary>
/// Shared response contract for About endpoints across all apps (React host, API, MVC).
/// Every About response is a single tree of these nodes, with the first child of any
/// root always being the app's own node, followed by any dependency nodes.
/// </summary>
public class AboutNode
{
    public required string Name { get; set; }
    public string? PublicMessage { get; set; }
    public bool IsHealthy { get; set; } = true;
    public string? Version { get; set; }
    public DateTime? BuildStart { get; set; }
    public int? BuildNumber { get; set; }
    public string? BuildBranchName { get; set; }
    public List<AboutNode> Children { get; set; } = new();
}

File: ui-react/src/about/aboutNode.js
/**
 * Shared About response contract used by every About endpoint (React host, API, MVC).
 * Each node: { name, isHealthy, version, buildStart, buildNumber, children }.
 */

/**
 * Creates a leaf node representing an app or dependency itself.
 * Version/build metadata is intentionally left null for now (real CI/CD metadata is out of scope).
 */
export function createSelfNode(name, isHealthy = true) {
  return {
    name,
    isHealthy,
    version: null,
    buildStart: null,
    buildNumber: null,
    children: [],
  };
}

/**
 * Creates a root node. The first entry in `children` must be the self node for that app,
 * followed by any dependency subtrees, in deterministic order.
 * Root health is the aggregate of all descendant health (all true => true).
 */
export function createRootNode(name, children) {
  return {
    name,
    isHealthy: children.every(isSubtreeHealthy),
    version: null,
    buildStart: null,
    buildNumber: null,
    children,
  };
}

export function isSubtreeHealthy(node) {
  return Boolean(node) && node.isHealthy && (node.children ?? []).every(isSubtreeHealthy);
}
